src/LuzFaltex.Extensions.DependencyInjection/ServiceLookup/CallSiteFactory.cs

```diff
@@ -0,0 +1,696 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Xml.Linq;
+using LuzFaltex.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup;
+
+[RequiresDynamicCode(MutableServiceProvider.RequiresDynamicCodeMessage)]
+internal sealed class CallSiteFactory : IServiceProviderIsService
+{
+    private bool DefaultCollectionChecked;
+    private const int DefaultSlot = 0;
+    private readonly Dictionary<string, IServiceCollection> _serviceCollections = new Dictionary<string, IServiceCollection>();
+    // private readonly ServiceDescriptor[] _descriptors;
+    private readonly ConcurrentDictionary<ServiceCacheKey, ServiceCallSite> _callSiteCache = new ConcurrentDictionary<ServiceCacheKey, ServiceCallSite>();
+    private readonly Dictionary<Type, ServiceDescriptorCacheItem> _descriptorLookup = new Dictionary<Type, ServiceDescriptorCacheItem>();
+    private readonly ConcurrentDictionary<Type, object> _callSiteLocks = new ConcurrentDictionary<Type, object>();
+
+    private readonly StackGuard _stackGuard;
+
+    public CallSiteFactory()
+    {
+        _stackGuard = new StackGuard();
+    }
+
+    internal void AddServiceCollection(string name, IServiceCollection serviceCollection)
+    {
+        _serviceCollections.Add(name, serviceCollection);
+        Populate();
```

Comment:
When using https://github.com/Foxtrek64/Remora.Plugins/pull/1 and debugging Remora.Plugins with a Discord Bot, somehow after having the services populated properly (before loading plugins), after plugins are loaded it can no longer find those services then (which can result in errors while loading other plugins).